### DISCUSSION TITLE
[Feature Request] Maintenance Mode

### DISCUSSION THREAD

**Discussion by colstrom** [runs-on/runs-on#170] [Ideas]
I'm not sure that the title is a good description, feel free to change it if you have a better name for it. 😆 

It would be nice to have a way to prepare for maintenance activities (such as upgrading RunsOn itself) in a non-disruptive manner, allowing **active jobs** to finish gracefully while preventing additional instances from being launched to pick up new jobs. Those jobs would just remain queued until the RunsOn deployment is taken out of this mode.

A simple version of this would just be a deploy-time parameter. Set it, app gets redeployed in that mode, operator does whatever activities are needed, and updates the CF stack to return to normal mode.

Alternatively, having this able to be defined in the repository configuration (perhaps a `disabled: true` option in `.github/runs-on.yml`?) , AND having a deploy-time parameter that designates a specific repository (defaulting to `.github-private` maybe?) to be used for org-wide control of this setting.

That approach would allow users to do whatever arbitrarily complex scheduling they want, without requiring RunsOn to know anything about it. It just reads the config, however it got that way is not its problem. User wants to do offline every night from 1800-2000? They can set up a GHA workflow with a cron trigger or something, etc.

Thanks for considering this feature. 🫶

**Comment by crohr** [comment]
@colstrom interesting idea, I'll move it to the discussions panel for now.

IMO, setting it from CF stack would be cumbersome and not immediate. Setting it from a repo config file could lead to edge cases if multiple edits in a short time, as well as low obersavability of the current state.

I'm wondering if we could simply use an SSM parameter set with the AWS CLI or via the UI instead? Since it is something that will be set by the operator, I guess it's fine if you need AWS access for that. That parameter could also be set through the AppRunner API endpoint with the `ServerPassword`, but probably better if it's not directly exposed.

**Reply by colstrom** [reply]
An SSM parameter could totally work, and would have a similar degree of flexibility. I proposed the config option simply because it seemed like an obvious-ish approach that probably already had most of what would be needed in the existing code, which I don't have access to (as far as I'm aware). So I was basically just guessing at what would be a low-effort implementation. 🤣

The important part is that it can be toggled with minimal disruption, and that in-flight jobs can finish as if nothing changed, while new jobs remain queued as if waiting for a worker to be available, rather than cancelled or rejected or some other form of disruption. Basically they'll still get worked _eventually,_ just... not **yet**.

I don't know that there's a way to provide some sort of status update to the queued jobs advising that things are in maintenance mode or whatever, but if there is, that would be a nice-to-have. I'm assuming that to be up to the operator to communicate to their users, though.
